### PR TITLE
feat: Status controls could have a way to have icons instead of the default letters

### LINKED ISSUE #1157: Feature Improvement: Status controls could have a way to have icons instead of the default letters
**Description**
For the Status GUI control, instead of showing E, S, I and W, it would be nice to have a way to put icons in it instead.

**Acceptance Criteria**
- [ ] Ensure new code is unit tested, and check code coverage is at least 90%
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change
- [ ] Ensure any change is well documented
- [ ] Status control from TaiPy gui can have Icons instead of the default letters

### PR DESCRIPTION
**Pull request by Vaibhav91one**
resolves #1157

### CODE REVIEW

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
Tx @Vaibhav91one Looks good 
need tests
what if
- the dev does not want icons but wants to keep the initials ?
- the dev wants to set its own icons ?

**Comment by Vaibhav91one** [comment]
> Tx @Vaibhav91one Looks good need tests what if
> 
> * the dev does not want icons but wants to keep the initials ?
> * the dev wants to set its own icons ?

So, the dev can use `getInitials `function which was being used before for the initials and if dev wants to set its own icons, just go to `getStatusIcon ` function and change the icon which are being chosen according to the switch case. @FredLL-Avaiga

**Comment by FredLL-Avaiga** [comment]
> > Tx @Vaibhav91one Looks good need tests what if
> > 
> > * the dev does not want icons but wants to keep the initials ?
> > * the dev wants to set its own icons ?
> 
> So, the dev can use `getInitials `function which was being used before for the initials and if dev wants to set its own icons, just go to `getStatusIcon ` function and change the icon which are being chosen according to the switch case. @FredLL-Avaiga

I suppose I was not clear enough.
The "dev" in this case is the person who writes a Taipy Application (he has no knowledge of js/ts or even how this works).
He just wants to show a status by writing some python/markdown that use the taipy components.

**Comment by Vaibhav91one** [comment]
> > > Tx @Vaibhav91one Looks good need tests what if
> > > 
> > > * the dev does not want icons but wants to keep the initials ?
> > > * the dev wants to set its own icons ?
> > 
> > 
> > So, the dev can use `getInitials `function which was being used before for the initials and if dev wants to set its own icons, just go to `getStatusIcon ` function and change the icon which are being chosen according to the switch case. @FredLL-Avaiga
> 
> I suppose I was not clear enough. The "dev" in this case is the person who writes a Taipy Application (he has no knowledge of js/ts or even how this works. He just wants to show a status by writing some python/markdown that use the taipy components.

Oh, I thought a developer from taipy team, mb. So, for the `intials` ig a different component would be better and if we are talking about setting `custom icons`, rn there is no way to do that as the icons are according to status.

**Comment by FredLL-Avaiga** [comment]
> 
> Oh, I thought a developer from taipy team, mb. So, for the `intials` ig a different component would be better and if we are talking about setting `custom icons`, rn there is no way to do that as the icons are according to status.

- A simple property could be used to define with_icons or not
- svg files could be provided for each type of status ...

**Comment by Vaibhav91one** [comment]
> > Oh, I thought a developer from taipy team, mb. So, for the `intials` ig a different component would be better and if we are talking about setting `custom icons`, rn there is no way to do that as the icons are according to status.
> 
> * A simple property could be used to define with_icons or not
> * svg files could be provided for each type of status ...

Hmm, Got it will look into it and update the PR.

**Comment by Vaibhav91one** [comment]
Hey @FredLL-Avaiga , I was adding the requested changes and require some guidance, please look into the following, 

In the statusProps I have added the following: 

```
interface StatusProps extends TaipyBaseProps {
    value: StatusType;
    onClose?: (evt: MouseEvent) => void;
    icon?: ReactNode; // Custom delete icon
    withIcons?: boolean; 
    customSvg?: ReactNode; 
}
```

And then the function for choosing whether it will be initials, icon or customSvg

```
const getStatusIcon = (status: string, withIcons?: boolean, customSvg?: ReactNode) => {
    const color = status2Color(status);
    const iconProps = { 
        sx: { fontSize: 20, color: `${color}.main` } // Customize icon size and color
    };

    if (customSvg) {
        return <div style={{ color: `${color}.main` }}>{customSvg}</div>; // Render the custom SVG
    }

    if (withIcons) {
        switch (status2Color(status)) {
            case "success":
                return <CheckCircleIcon {...iconProps} />;
            case "warning":
                return <WarningIcon {...iconProps} />;
            case "error":
                return <ErrorIcon {...iconProps} />;
            default:
                return <InfoIcon {...iconProps} />;
        }
    } else {
        return getInitials(status); 
    }
};
```

Later in status component, I am taking the props and using them,

```
const Status = (props: StatusProps) => {
    const { value, id, withIcons = false, customSvg } = props;

    const className = useClassNames(props.libClassName, props.dynamicClassName, props.className);


    const chipProps = useMemo(() => {
        const cp: Record<string, unknown> = {};
        const statusColor = status2Color(value.status);
        cp.color = statusColor;
        cp.avatar = (
            <Avatar 
                sx={{ 
                    bgcolor: "transparent",
                    color: `${statusColor}.main`
                }}
            >
                {getStatusIcon(value.status, withIcons, customSvg)}
            </Avatar>
        );
        if (props.onClose) {
            cp.onDelete = props.onClose;
        }
        if (props.icon) {
            cp.deleteIcon = props.icon;
        }
        return cp;
    }, [value.status, props.onClose, props.icon, withIcons, customSvg]);

    return <Chip id={id} variant="outlined" {...chipProps} label={value.message} sx={chipSx} className={className} />;
};
```
My py script is the following: 

```
from taipy.gui import Gui
from taipy.gui.icon import Icon

status = [
    ("warning", "Task is launched."),
    ("warning", "Taks is waiting."),
    ("error", "Task timeout."),
    ("info", "Process was cancelled.")
]

without_close = True
withIcons = True
icon = Icon("./halloween-clown.png", "Charles")

page = """
<|{status}|status|{withIcons}|customSvg={icon}|>
"""

if __name__ == "__main__":
    Gui(page).run(title="Status")
```

The issue here is I am not able to read the props being passed, I don't seem to get what I am doing wrong.

**Comment by FredLL-Avaiga** [comment]
Why don't you update your PR with this changes so that we can discuss on something concrete ?
You'll have to define new properties in the status declaration block in https://github.com/Avaiga/taipy/blob/3621961930cec46fd204c2322ff694542b0d85c9/taipy/gui/_renderers/factory.py#L543

you cannot pass a ReactNode typed property, but you can pass an URL or a svg string, like it's done in the image component.
https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/Image.tsx#L55

**Comment by Vaibhav91one** [comment]
> Why don't you update your PR with this changes so that we can discuss on something concrete ? You'll have to define new properties in the status declaration block in
> 
> https://github.com/Avaiga/taipy/blob/3621961930cec46fd204c2322ff694542b0d85c9/taipy/gui/_renderers/factory.py#L543
> 
> you cannot pass a ReactNode typed property, but you can pass an URL or a svg string, like it's done in the image component.
> 
> https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/Image.tsx#L55

Will do this and update the PR.

**Comment by Vaibhav91one** [comment]
> > Why don't you update your PR with this changes so that we can discuss on something concrete ? You'll have to define new properties in the status declaration block in
> > https://github.com/Avaiga/taipy/blob/3621961930cec46fd204c2322ff694542b0d85c9/taipy/gui/_renderers/factory.py#L543
> > 
> > you cannot pass a ReactNode typed property, but you can pass an URL or a svg string, like it's done in the image component.
> > https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/Image.tsx#L55
> 
> Will do this and update the PR.

Hi @FredLL-Avaiga , sorry for disturbing again, but after making changes in factory.py, do I need to do anything else besides running **npm run build:dev** in taipy-gui. Also, I tried to run factory.py but it showed the following error.

```
    from ..types import PropertyType
ImportError: attempted relative import with no known parent package
```

**Comment by FredLL-Avaiga** [comment]
> > > Why don't you update your PR with this changes so that we can discuss on something concrete ? You'll have to define new properties in the status declaration block in
> > > https://github.com/Avaiga/taipy/blob/3621961930cec46fd204c2322ff694542b0d85c9/taipy/gui/_renderers/factory.py#L543
> > > 
> > > you cannot pass a ReactNode typed property, but you can pass an URL or a svg string, like it's done in the image component.
> > > https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/Image.tsx#L55
> > 
> > 
> > Will do this and update the PR.
> 
> Hi @FredLL-Avaiga , sorry for disturbing again, but after making changes in factory.py, do I need to do anything else besides running **npm run build:dev** in taipy-gui. Also, I tried to run factory.py but it showed the following error.
> 
> ```
>     from ..types import PropertyType
> ImportError: attempted relative import with no known parent package
> ```

You need to create a python script and run it to show your component on a page
there are examples in doc/gui/examples

**Comment by Vaibhav91one** [comment]
> > > > Why don't you update your PR with this changes so that we can discuss on something concrete ? You'll have to define new properties in the status declaration block in
> > > > https://github.com/Avaiga/taipy/blob/3621961930cec46fd204c2322ff694542b0d85c9/taipy/gui/_renderers/factory.py#L543
> > > > 
> > > > you cannot pass a ReactNode typed property, but you can pass an URL or a svg string, like it's done in the image component.
> > > > https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/Image.tsx#L55
> > > 
> > > 
> > > Will do this and update the PR.
> > 
> > 
> > Hi @FredLL-Avaiga , sorry for disturbing again, but after making changes in factory.py, do I need to do anything else besides running **npm run build:dev** in taipy-gui. Also, I tried to run factory.py but it showed the following error.
> > ```
> >     from ..types import PropertyType
> > ImportError: attempted relative import with no known parent package
> > ```
> 
> You need to create a python script and run it to show your component on a page there are examples in doc/gui/examples

Oh, that's clear to me, the python script is working. The only thing I am not understanding is the property. As you guided I created a `with_icons` property same as `without_close` which will be False by default and similar to without_close I am using that in python script but the Status component is showing undefined for the prop which is `withIcons`.

**taipy/taipy/gui/_renderers/factory.py**:
```
 .set_attributes(
            [
                ("without_close", PropertyType.boolean, False),
                ("with_icons", PropertyType.boolean, False),
                ("hover_text", PropertyType.dynamic_string),
            ]
        ),
```

**Python Script**:

```
with_icons = True
content = "./halloween-clown.png"

page = """
<|{status}|status|{with_icons}|>
<|{content}|image|label=This is an image|on_action=function_name|>
"""

if __name__ == "__main__":
    Gui(page).run(title="Status")
```

**Status Component:**

```
interface StatusProps extends TaipyBaseProps {
    value: StatusType;
    onClose?: (evt: MouseEvent) => void;
    icon?: ReactNode; // Custom delete icon
    withIcons?: boolean; 
    content?: string; 
}
```

**Comment by FredLL-Avaiga** [comment]
> <|{status}|status|{with_icons}|>

`<|{status}|status|with_icons|>` or `<|{status}|status|with_icons=True|>`

**Comment by Vaibhav91one** [comment]
> > <|{status}|status|{with_icons}|>
> 
> `<|{status}|status|with_icons|>` or `<|{status}|status|with_icons=True|>`

Tried the above, still not working.

**Comment by FredLL-Avaiga** [comment]
* Here you see the frontend component
https://github.com/Avaiga/taipy/blob/3621961930cec46fd204c2322ff694542b0d85c9/taipy/gui/_renderers/factory.py#L538
* Here you can find the front end association
https://github.com/Avaiga/taipy/blob/3621961930cec46fd204c2322ff694542b0d85c9/frontend/taipy-gui/src/components/Taipy/index.ts#L79
* That means that you need to add your `withIcons` property to the `StatusList` Component
https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/StatusList.tsx#L88
* and forward it to the `Status` component
https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/StatusList.tsx#L162
https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/StatusList.tsx#L171

**Comment by Vaibhav91one** [comment]
> * Here you see the frontend component
>   https://github.com/Avaiga/taipy/blob/3621961930cec46fd204c2322ff694542b0d85c9/taipy/gui/_renderers/factory.py#L538
> * Here you can find the front end association
>   https://github.com/Avaiga/taipy/blob/3621961930cec46fd204c2322ff694542b0d85c9/frontend/taipy-gui/src/components/Taipy/index.ts#L79
> * That means that you need to add your `withIcons` property to the `StatusList` Component
>   https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/StatusList.tsx#L88
> * and forward it to the `Status` component
>   https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/StatusList.tsx#L162
>   
>   https://github.com/Avaiga/taipy/blob/8379661f1380a7722f1643b1252f62df55b5df8d/frontend/taipy-gui/src/components/Taipy/StatusList.tsx#L171

Oh, will do that. Thanks @FredLL-Avaiga.

**Comment by Vaibhav91one** [comment]
Hi @FredLL-Avaiga, did the requested changes PFA, screenshots and python script for testing.

1. **Default behaviour** : 
default

2. **`with_icons` property**:
with_icons

3. **`custom_icon` property**:
custom_icon works with single svg also
CustomIcons


**Python Script for debugging:**

```
from taipy.gui import Gui

status = [
    ("warning", "Task is launched."),
    ("warning", "Taks is waiting."),
    ("error", "Task timeout."),
    ("info", "Process was cancelled.")
]

with_icons = True

// For SVG url
custom_icon = "https://www.svgrepo.com/show/530594/pants.svg;https://www.svgrepo.com/show/530595/hotel.svg;https://www.svgrepo.com/show/530596/diving-goggles.svg;https://www.svgrepo.com/show/530597/hat.svg;https://www.svgrepo.com/show/530598/building.svg"

// for Local SVG Icons
custom_icon = "./mickey-and-goofy-halloween.svg;./mickey-and-goofy-halloween.svg;./mickey-and-goofy-halloween.svg;./mickey-and-goofy-halloween.svg;./mickey-and-goofy-halloween.svg"



page = """
<|{status}|status|{custom_icon}|>
<|{status}|status|{with_icons}|>
<|{status}|status|>
"""

if __name__ == "__main__":
    Gui(page).run(title="Status")
    

```

File: frontend/taipy-gui/src/components/Taipy/StatusList.tsx [hunk 1]
```diff
@@ -95,6 +97,16 @@ const StatusList = (props: StatusListProps) => {
     const [opened, setOpened] = useState(false);
     const [multiple, setMultiple] = useState(false);
     const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);
+    const content = useMemo(() => {
+        if (props.customIcon) {
+            try {
+                return props.customIcon.split(';');
+            } catch (e) {
+                console.info(`Error parsing custom icons\n${(e as Error).message || e}`);
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/StatusList.tsx:105]: How could this throw an exception except if this is not a string ?
Better test the type and ignore if it is not a string

File: frontend/taipy-gui/src/components/Taipy/Status.tsx [hunk 2]
```diff
@@ -44,25 +50,104 @@ const status2Color = (status: string): "error" | "info" | "success" | "warning"
     return "info";
 };
 
+// Function to get the appropriate icon based on the status
+const GetStatusIcon = (status: string, withIcons?: boolean): ReactNode => {
+
+    const color = status2Color(status);
+    const iconProps = {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Status.tsx:59]: useMemo in the main component

File: frontend/taipy-gui/src/components/Taipy/Status.tsx [hunk 3]
```diff
@@ -44,25 +50,104 @@ const status2Color = (status: string): "error" | "info" | "success" | "warning"
...
 const chipSx = { alignSelf: "flex-start" };
 
 const Status = (props: StatusProps) => {
     const { value, id } = props;
-
+    const content = props.content;
+    const withIcons = props.withIcons;
     const className = useClassNames(props.libClassName, props.dynamicClassName, props.className);
 
+    const isSvgUrl = (content?: string) => {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Status.tsx]: define outside the component

File: frontend/taipy-gui/src/components/Taipy/Status.tsx [hunk 4]
```diff
@@ -44,25 +50,104 @@ const status2Color = (status: string): "error" | "info" | "success" | "warning"
...
+            return content?.substring(content?.length - 4).toLowerCase() === ".svg"; // Check if it ends with ".svg"
+    };
+
+    const isInlineSvg = (content?: string) => {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Status.tsx]: define  outside the component

File: frontend/taipy-gui/src/components/Taipy/Status.tsx [hunk 5]
```diff
@@ -44,25 +50,104 @@ const status2Color = (status: string): "error" | "info" | "success" | "warning"
...
+        if (isSvgUrl(content)) {
+            cp.avatar = (
+                <Avatar src={content} />
+            );
+        } 
+        
+        else if(content && isInlineSvg(content)){
+            cp.avatar = (
+                <Avatar
+                    sx={{
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Status.tsx]: const outside the component

File: frontend/taipy-gui/src/components/Taipy/Status.tsx [hunk 6]
```diff
@@ -44,25 +50,104 @@ const status2Color = (status: string): "error" | "info" | "success" | "warning"
...
+                        bgcolor: "transparent"
+                    }}
+                >
+                    <div
+                      style={{
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Status.tsx]: const outside the component

File: frontend/taipy-gui/src/components/Taipy/Status.tsx [hunk 7]
```diff
@@ -44,25 +50,104 @@ const status2Color = (status: string): "error" | "info" | "success" | "warning"
...
+                        width: '100%',
+                        height: '100%',
+                        display: 'flex',
+                        alignItems: 'center',
+                        justifyContent: 'center',
+                    }}
+                    dangerouslySetInnerHTML={{ __html: content }} />
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Status.tsx]: any other way ?

File: frontend/taipy-gui/src/components/Taipy/Status.tsx [hunk 8]
```diff
@@ -44,25 +50,104 @@ const status2Color = (status: string): "error" | "info" | "success" | "warning"
...
+                </Avatar>
+            );
+        }
+
+        else {
+            cp.avatar = (
+                <Avatar
+                    sx={{
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Status.tsx:149]: useMemo

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
Can you add your example in the doc/gui/example directory ?
And one or 2 frontend tests ?

**Comment by Vaibhav91one** [comment]
> Can you add your example in the doc/gui/example directory ? And one or 2 frontend tests ?

Yup, will do that and the requested changes. Regarding the frontend tests what do you mean exactly? Here, we are talking about my python script or jest files for testing. If jest file, Can you please point towards an existing test file for context. @FredLL-Avaiga

**Comment by FredLL-Avaiga** [comment]
> > Can you add your example in the doc/gui/example directory ? And one or 2 frontend tests ?
> 
> Yup, will do that and the requested changes. Regarding the frontend tests what do you mean exactly? Here, we are talking about my python script or jest files for testing. If jest file, Can you please point towards an existing test file for context. @FredLL-Avaiga

https://github.com/Avaiga/taipy/blob/569f29cc06887726e85ba98c428ee7e4e825303c/frontend/taipy-gui/src/components/Taipy/Status.spec.tsx#L25

**Comment by FredLL-Avaiga** [comment]
can you fix the linter issues ?

**Comment by Vaibhav91one** [comment]
> can you fix the linter issues ?

Yeah will do that, It was in `factory.py` file right?

**Comment by Vaibhav91one** [comment]
@FredLL-Avaiga  Did the requested changes and added few test cases.

File: frontend/taipy-gui/src/components/Taipy/Status.tsx (hunk 8 above)
Comment (Vaibhav91one) [reply frontend/taipy-gui/src/components/Taipy/Status.tsx:149]: chip.props is already being memoized here.

**Comment by FredLL-Avaiga** [comment]
front end test fails

**Comment by Vaibhav91one** [comment]
> front end test fails

Hi @FredLL-Avaiga , it was `useMemo` Hooks that were causing the issue. All Tests passed for StatusList component.

**Review by FredLL-Avaiga — APPROVED** [review]

**Comment by jrobinAV** [comment]
Thank you @Vaibhav91one !

**Comment by Vaibhav91one** [comment]
Thank you @jrobinAV and @FredLL-Avaiga thank you for the guidance, learned a lot.

**Comment by RymMichaut** [comment]
hey @Vaibhav91one congratulations!
Please reach out to me at rym.michaut@taipy.io to prepare your swag shipping

**Pull request merged**